### ISSUE TITLE
Run processes inside container as non-root user

### ISSUE THREAD

**Issue by graduatio** [TandoorRecipes/recipes#1257]
**Is your feature request related to a problem? Please describe.**
The processes inside the recipes container currently run as user root. This is potentially problematic for security reasons.

**Describe the solution you'd like**
Whenever possible the processes should run as non-root user.

**Comment by graduatio** [comment]
I tried to run all containers (nginx, postgres and recipes) as a non-root user. The nginx and postgres docker documentations describe how the containers can run as an arbitrary non-privileged user. I followed the steps below:

For the nginx container:

- added `user: "${UID}:${GID}"` in `docker-compose.yaml` to the `nginx_recipes` section
- added to the `volumes` section in `docker-compose.yaml`:
	`- ./nginx.conf:/etc/nginx/nginx.conf`
- created a `nginx.conf` file in the current directory with following redefined directives:
```
pid        /tmp/nginx.pid;
http {
    client_body_temp_path /tmp/client_temp;
    proxy_temp_path       /tmp/proxy_temp_path;
    fastcgi_temp_path     /tmp/fastcgi_temp;
    uwsgi_temp_path       /tmp/uwsgi_temp;
    scgi_temp_path        /tmp/scgi_temp;
}
```
For the postgres container:
- changed to a Debian Variant image (e.g. postgres:latest)
- added `user: "${UID}:${GID}"` in `docker-compose.yaml` to the `db_recipes` section

For the recipe container:
- added `user: "${UID}:${GID}"` in `docker-compose.yaml` to the `web_recipes` section

Before I started the containers:
- added `UID=` and `GID=` with the desired values to the `.env` file
- created a `mediafiles` and a `postgresql` directory and used `chown` to change ownership to the user which should run the containers

With this setup the database and the nginx server seem to work properly. Tandoor can be accessed by opening the ip-address. But the docker-compose logs show permission errors of the recipes container which cannot change files in `/opt/recipes` when executing the commands in `boot.sh`.
This error occures because `opt/recipes` was created by user root. Changing the owner of the files inside the container manually resolves the problem temporarily.
Maybe this issue can be resolved by setting a default user and changing the owner of `opt/recipes` in the Dockerfile.

**Comment by vabene1111** [comment]
thanks, i will take a look at this at some point when the more pressing features are taken care of

**Comment by Akruidenberg** [comment]
Any progress so far. I would like to run the recipe container rootless. The trick of @graduatio is working, only manually .

**Comment by graduatio** [comment]
@Akruidenberg unfortunately the processes in the default container still run as user root. To simplify the modification process of the recipe container to run as a non-root user I currently use following docker-compose.yaml in combination with an additional Dockerfile in `./recipes`:


**docker-compose.yaml**
```
version: "3"
services:
  db_recipes:
    restart: always
    image: postgres:alpine
    volumes:
      - ./postgresql:/var/lib/postgresql/data
    env_file:
      - ./.env

  web_recipes:
    build: ./recipes/
    user: "${UID}:${GID}"
    restart: always
    env_file:
      - ./.env
    volumes:
      - staticfiles:/opt/recipes/staticfiles
      - nginx_config:/opt/recipes/nginx/conf.d
      - ./mediafiles:/opt/recipes/mediafiles
    depends_on:
      - db_recipes

  nginx_recipes:
    image: nginx:mainline-alpine
    restart: always
    ports:
      - 80:80
    env_file:
      - ./.env
    depends_on:
      - web_recipes
    volumes:
      - nginx_config:/etc/nginx/conf.d:ro
      - staticfiles:/static:ro
      - ./mediafiles:/media:ro

volumes:
  nginx_config:
  staticfiles:
```


**Dockerfile**
```
FROM vabene1111/recipes:latest
RUN adduser -D --uid 999 tandoor
RUN chown -R 999:999 /opt/recipes
USER 999:999
ENTRYPOINT ["/opt/recipes/boot.sh"]
```

The`UID` and `GID` have to be specified in the `.env` file and should match the values in the Dockerfile.

**Comment by Akruidenberg** [comment]
@graduatio thanks. Will try it out

**Comment by dsm1212** [comment]
Any update on this? With the advent of rootless docker, most containers support setting user now through compose or env vars. This is the only one I have left that doesn't support it.

**Comment by smilerz** [comment]
Feel free to submit a PR with an updated Dockerfile.

**Comment by dsm1212** [comment]
Ok, so the issue really is that during startup django copies some files from the staticfiles volume to /opt/recipes/cookbook/static. For me it was just one file (django_js_reverse/reverse.js). But that tree is all owned by root so if the container is started with a user then that copy fails. Aside from that you just have to make sure that all volumes are chown to the user you are using. Is there a point in the build where this static folder is created we could just change the default permissions inside the image on /opt/recipes/cookbook/static/*? I don't know enough about what is going on to know if it is always just this one file or could be the whole tree.

**Comment by rettops** [comment]
What worked for me is to map the cookbook/static directory along with the others in my docker-compose.yml:
```
    volumes:
      - ./recipes_data/staticfiles:/opt/recipes/staticfiles
      - ./recipes_data/cookbook/static:/opt/recipes/cookbook/static
      - ./recipes_data/mediafiles:/opt/recipes/mediafiles
```

**Comment by dsm1212** [comment]
> What worked for me is to map the cookbook/static directory along with the others in my docker-compose.yml:
> 
> ```
>     volumes:
>      ...
>       - ./recipes_data/cookbook/static:/opt/recipes/cookbook/static

I tried this and cookbook/static just ended up containing the django_js_reversed/reverse.js file.  But this cookbook/static folder has a ton of contents in the image and the app doesn't work without them. Did you copy the files out of the image? That would make updating hard. However, I did find that if I mapped this subfolder which contains the one file that gets copied it works though, so thanks for the suggestion:

 - ./recipes_data/cookbook/static/django_js_reversed:/opt/recipes/cookbook/static/django_js_reversed

**Comment by rettops** [comment]
Hmm. You're right, my cookbook/static directory ended up containing only
the django_js_reverse directory, but the app seemed to work fine. In any
case, your solution is probably better, because it's more surgical. I think
you may have a typo, because my copy of the app expects a django_js_reverse
directory rather than django_js_reversed.

      -
> ./recipes_data/cookbook/static/django_js_reverse:/opt/recipes/cookbook/static/django_js_reverse

**Comment by dsm1212** [comment]
Keeping this working has been a pain so I'm running as root again. Could we get this looked into? The biggest thorn with the current image is that when it runs nginx it runs as the "user" specified in compose, but the image has folders owned and only accessible by nginx. Also, the socket created in /run isn't allowed from a normal user. I think most of the issues could be addressed by relaxing permissions on a few configu folders internal to the container, but the real nginx contain image has support for user so maybe it would be better to look at what they did there.

**Comment by dsm1212** [comment]
Ah, well, I stared at the current problem some more. I was still using the three container config from above I switched to the default two container one and it got easier. The django reverse file override seems no longer necessary, although I don't recall the circumstance when that mattered so maybe it will bite me to take out. It appears I only needed these two:
      # map paths to temp, this is standard howto from the nginx site
      - /apps/tandoor/http.d:/opt/recipes/http.d
      # map run to a folder with more open permissions so a socket can be created
      - /apps/tandoor/run:/run
Which is not so bad, but could the dockerfile set up a subfolder owned by the specified user?

**Comment by vabene1111** [comment]
sorry but I already lack the time to work on tandoor properly, I simply cannot support non root setups for the time being. If someone provides a credible source that running as root is actually dangerous I will look into this more but until then I cant justify spending so much time on a nice to have feature.

**Comment by dsm1212** [comment]
It's ok, these two things are not so bad. I had been running the three image config someone shared above earlier and that was getting pretty confusing.

The problem with running as root is mostly the risk of malicious code injection. When we pull open source container updates, one could be poisoned with something nefarious that got slipped in. It's unlikely but there are a lot of layers that it could happen so it's nearly impossible for any single project to know they are safe.

**Comment by vabene1111** [comment]
I do not understand that argument. Containers are meant, besides other things, to protect the host system from malicious action within the container. You obviously should not mount any security relevant host directories into the container but besides that I do not see how configuring the container to run and non root would protect you from malicious code I or someone else adds to the repo. 

The only thing I can see is lower privileges for runtime attacks on the server but even then its still a container so its decoupled from the host system.

**Comment by tribut** [comment]
Its about all about risk management. Like all software, Docker is not without (security) bugs and some of these allow escalation to the host or other containers.
Running with minimal privileges drastically reduces the threat surface for any potential such vulnerability.

See for example this OWASP cheat sheet that lists the most important consideration for container security:
https://cheatsheetseries.owasp.org/cheatsheets/Docker_Security_Cheat_Sheet.html
Rule 2 is not running as root.

So in that way it is "nice to have" as all security features are.

**Comment by vabene1111** [comment]
ok, thank you, I can accept an OWASP recommendation. Since there is such broad intrest from the community on improving this I would appreciate help in making this change as I am not a huge expert on all the build files and related stuff.

**Comment by tribut** [comment]
If nobody else feels like it, I will look into it probably sometime next week.

**Comment by dsm1212** [comment]
FWIW the postgres container supports setting user/group via the compose standard options. But if you have created the database already as root to switch to another user you have to first "chmod -R" the entire database volume.

RIght now the problem I have with the web image is two things:

1. nginx needs to remap a list of root owned folders to /tmp. That is discussed above and if you search the nginix docker image that is the proscribed way around it. Since you are embedding nginx maybe you just permanently do that in the image by adding this file to http.d:

# cat nginx_user.conf
    client_body_temp_path /tmp/client_temp;
    proxy_temp_path       /tmp/proxy_temp_path;
    fastcgi_temp_path     /tmp/fastcgi_temp;
    uwsgi_temp_path       /tmp/uwsgi_temp;
    scgi_temp_path        /tmp/scgi_temp;

2. The web image tries to create /run/tandoor.socket and fails. To work around I mapped a volume to the /run folder with an nginx subfolder, all owned by my tandoor user. I think if the dockerfile were to create /run/tandoor and make it owned by the specified user and then the socket was created in that subfolder it would fix the problem.

**Comment by sdwilsh** [comment]
I gave it a go to run this as non-root and read-only image file system (technically scope creep for this issue, but it's [Rule #8 from OWASP](https://cheatsheetseries.owasp.org/cheatsheets/Docker_Security_Cheat_Sheet.html#rule-8-set-filesystem-and-volumes-to-read-only)) in kubernetes tonight, and there were a few problems that I noticed based on these relevant log lines:
```
/opt/recipes/boot.sh: line 21: can't create /opt/recipes/http.d/Recipes.conf: Read-only file system
Starting nginx
nginx: [alert] could not open error log file: open() "/var/lib/nginx/logs/error.log" failed (13: Permission denied)
2026/01/29 22:09:06 [warn] 9#9: the "user" directive makes sense only if the master process runs with super-user privileges, ignored in /etc/nginx/nginx.conf:3
2026/01/29 22:09:06 [emerg] 10#10: open() "/run/nginx/nginx.pid" failed (30: Read-only file system)
nginx: [emerg] open() "/run/nginx/nginx.pid" failed (30: Read-only file system)
....
chmod: /opt/recipes/mediafiles: Operation not permitted
chmod: /opt/recipes/mediafiles: Operation not permitted
Starting gunicorn
[2026-01-29 22:09:37 -0800] [7] [INFO] Starting gunicorn 23.0.0
[2026-01-29 22:09:37 -0800] [7] [ERROR] connection to /run/tandoor.sock failed: [Errno 30] Read-only file system
[2026-01-29 22:09:38 -0800] [7] [ERROR] connection to /run/tandoor.sock failed: [Errno 30] Read-only file system
[2026-01-29 22:09:39 -0800] [7] [ERROR] connection to /run/tandoor.sock failed: [Errno 30] Read-only file system
[2026-01-29 22:09:40 -0800] [7] [ERROR] connection to /run/tandoor.sock failed: [Errno 30] Read-only file system
[2026-01-29 22:09:41 -0800] [7] [ERROR] connection to /run/tandoor.sock failed: [Errno 30] Read-only file system
[2026-01-29 22:09:42 -0800] [7] [ERROR] Can't connect to /run/tandoor.sock
```
- `boot.sh` carries on despite commands erroring.  We probably want to `set -e` at the top of that file so it fails after an error.
- `/opt/recipes/http.d/Recipes.conf` can't be written out due to read-only filesystem.  I can work around this by mounting a tmpfs volume there.
- nginx error and pid files cannot be written out.  The pid file has the same issue as gunicorn (see below) although it [could be configured to be elsewhere](https://nginx.org/en/docs/ngx_core_module.html#pid), and the error log can be worked around with an error log, although it'd probably be better to somehow surface those logs into the container's stderr so that they are easy to see from docker/podman/kubectl
- unconditional chmod is not ideal if you drop all privledges.  I think this is probably not needed if the files have the right permissions before being copied
- the socket file to use with gunicorn could be controllable with an env variable, or possibly the suggestion by @dsm1212 would work as well since we have the same issue with nginx.

**Comment by sdwilsh** [comment]
It actually looks like @wilmardo has done pretty much all of the work to run as non-root for this in #4042.  The read-only filesystem stuff wouldn't be too hard to fix once that is merged.

**Comment by wilmardo** [comment]
Yes feel free to try that branch, I have just updated it for 2.4.0 and it have it running for a couple months now.

**Comment by vabene1111** [comment]
thank you very much, still planning to merge this, I just had some other priorities and very limited time.

**Comment by NotesTaker** [comment]
@sdwilsh Did you succeed with running it rootless? 
Trying with latest version but I do not manage to get it connected when using a non-root user (all is fine with user root). Container is starting and running but I do not get any connection at the assigned port. The container log does not show any error messages.

**Comment by dsm1212** [comment]
So I cannot remember the sequence of events, but I'm running tandoor latest as non-root. I can see I removed my earlier work-arounds. I don't recall how it is working given what appears to be a lack of merge to release. I must have looked at latest and saw the fix was there or something.  This is my compose, ports are left to default bcause I have swag in this network and it's the only way in.

`services:
  db_recipes:
    hostname: db_recipes
    image: postgres:15-alpine
    container_name: tandoor_db
    network_mode: svcnet
    env_file: ./tandoor.env
    user: 1032:100
    volumes:
      - /apps/tandoor/postgresql-15:/var/lib/postgresql/data
      - /etc/passwd:/etc/passwd:ro                               #SKIP_BACKUP
    healthcheck:
      test: ["CMD-SHELL", "pg_isready -d djangodb -U djangouser"]
      interval: 30s
      start_period: 30s
      timeout: 10s
      retries: 5
    restart: unless-stopped

  web_recipes:
    hostname: web_recipes
    image: vabene1111/recipes:latest
    #build: ./recipes/
    container_name: tandoor_web
    network_mode: svcnet
    env_file: ./tandoor.env
    user: 1032:100
    volumes:
      - /apps/tandoor/staticfiles:/opt/recipes/staticfiles                    #SKIP_BACKUP
      - /apps/tandoor/mediafiles:/opt/recipes/mediafiles                      #SKIP_BACKUP
    depends_on:
      db_recipes:
        condition: service_healthy
    restart: unless-stopped`

and ps -ef | grep tandoor:

`root@warehouse21:/apps/tandoor# ps -ef | grep tandoor
tandoor     3083    3045  0 Jul26 ?        00:00:08 /sbin/tini -- /opt/recipes/boot.sh

tandoor     3171    3102  0 Jul26 ?        00:00:10 postgres

tandoor     3757    3083  0 Jul26 ?        00:00:41 /opt/recipes/venv/bin/python /opt/recipes/venv/bin/gunicorn --bind unix:/tmp/tandoor.sock --workers 3 --threads 2 --timeout 30 --access-logfile - --error-logfile - --log-level warning recipes.wsgi

tandoor     3864    3083  0 Jul26 ?        00:00:00 nginx: master process nginx -g pid /tmp/nginx.pid;

tandoor     3865    3864  0 Jul26 ?        00:00:00 nginx: worker process

tandoor     3866    3864  0 Jul26 ?        00:00:00 nginx: worker process

tandoor     3867    3864  0 Jul26 ?        00:00:00 nginx: worker process

tandoor     3868    3864  0 Jul26 ?        00:00:00 nginx: worker process

tandoor     3869    3864  0 Jul26 ?        00:00:00 nginx: worker process

tandoor     3870    3864  0 Jul26 ?        00:00:00 nginx: worker process

tandoor     3871    3864  0 Jul26 ?        00:00:00 nginx: worker process

tandoor     3872    3864  0 Jul26 ?        00:00:00 nginx: worker process

tandoor     4373    3171  0 Jul26 ?        00:00:00 postgres: checkpointer

tandoor     4374    3171  0 Jul26 ?        00:00:01 postgres: background writer

tandoor     4510    3171  0 Jul26 ?        00:00:01 postgres: walwriter

tandoor     4512    3171  0 Jul26 ?        00:00:02 postgres: autovacuum launcher

tandoor     4513    3171  0 Jul26 ?        00:00:00 postgres: logical replication launcher

tandoor     9556    3757  0 Jul26 ?        00:00:49 /opt/recipes/venv/bin/python /opt/recipes/venv/bin/gunicorn --bind unix:/tmp/tandoor.sock --workers 3 --threads 2 --timeout 30 --access-logfile - --error-logfile - --log-level warning recipes.wsgi

tandoor     9557    3757  0 Jul26 ?        00:00:50 /opt/recipes/venv/bin/python /opt/recipes/venv/bin/gunicorn --bind unix:/tmp/tandoor.sock --workers 3 --threads 2 --timeout 30 --access-logfile - --error-logfile - --log-level warning recipes.wsgi

tandoor     9577    3757  0 Jul26 ?        00:00:50 /opt/recipes/venv/bin/python /opt/recipes/venv/bin/gunicorn --bind unix:/tmp/tandoor.sock --workers 3 --threads 2 --timeout 30 --access-logfile - --error-logfile - --log-level warning recipes.wsgi

tandoor  3669658    3171  0 20:17 ?        00:00:00 postgres: djangouser djangodb 172.18.0.6(38672) idle

tandoor  3669662    3171  0 20:17 ?        00:00:00 postgres: djangouser djangodb 172.18.0.6(60304) idle

tandoor  3669686    3171  0 20:17 ?        00:00:00 postgres: djangouser djangodb 172.18.0.6(60306) idle

tandoor  3669687    3171  0 20:17 ?        00:00:00 postgres: djangouser djangodb 172.18.0.6(60316) idle

tandoor  3669688    3171  0 20:17 ?        00:00:00 postgres: djangouser djangodb 172.18.0.6(60326) idle

tandoor  3669689    3171  0 20:17 ?        00:00:00 postgres: djangouser djangodb 172.18.0.6(60332) idle`

**Comment by NotesTaker** [comment]
Thanks for your help. I reverted all my work arounds after your feeedback and indeed - then it worked like a charm. Only thing I needed to update was TANDOOR_PORT to use a port above 1024.

**Comment by wilmardo** [comment]
Probably because !4042 got shipped so since then the container can run as non-root but defaults to root for backwards compatibility. 

@NotesTaker @dsm1212 Thanks for posting in that things are working!

@graduatio have you tried again with the latest release?

I would be in favor of closing this issue :)
